shared/history.py

```diff
@@ -0,0 +1,91 @@
+    History.update_history: Updates the history with a new match, adding the match ID, state, action, and reward.
+    History.dump: Serializes the history of matches to a JSON string.
+"""
+
+from typing import List, Tuple, Annotated, Optional, Dict
+from pydantic import BaseModel
+from shared.random_player import RandomPlayer
+from shared.utils import AbstractPlayer, Action, State, Turn, Color
+
+class Match(BaseModel):
+    """
+    Model representing a match in the Tablut game.
+
+    Attributes:
+        match_id (int): The unique identifier for the match.
+        white_player (AbstractPlayer): The player playing as white.
+        black_player (AbstractPlayer): The player playing as black.
+        turns (List[Tuple[State, Action, float]]): The list of turns taken during the match, each containing a state, an action, and a reward.
+        outcome (Optional[Turn]): The outcome of the match.
+    """
+    match_id: Annotated[int, "The unique identifier for the match"]
+    white_player: Annotated[AbstractPlayer, "The player playing as white"]
+    black_player: Annotated[AbstractPlayer, "The player playing as black"]
+    turns: Annotated[List[Tuple[State, Action, float]], "The list of turns taken during the match, each containing a state, an action, and a reward"]
+    outcome: Annotated[Optional[Turn], "The outcome of the match"]
+
+    def __str__(self) -> str:
+        """
+        Returns a human-readable string representing the match.
+
+        Returns:
+            str: A string with match details.
+        """
+        turns_str = "\n".join(
+            f"Turn {i + 1}:\nState:\n{state}\nAction: {action}\nReward: {reward}"
+            for i, (state, action, reward) in enumerate(self.turns)
+        )
+        return (
+            f"Match ID: {self.match_id}\n"
+            f"White Player: {self.white_player.name}\n"
+            f"Black Player: {self.black_player.name}\n"
+            f"Turns:\n{turns_str}\n"
+            f"Outcome: {self.outcome}"
+        )
+
+class History(BaseModel):
+    """
+    Class representing the history of matches played.
+
+    Attributes:
+        matches (dict[int, Match]): A dictionary mapping match IDs to Match objects.
+    """
+    matches: Annotated[Dict[int, Match], "A dictionary mapping match IDs to Match objects"]
+
+    def update_history(self, match_id: int, state: State, action: Action, reward: float):
+        """
+        Updates the history with a new match, adding the match ID, state, action, and reward.
+
+        Args:
+            match_id (int): The unique identifier for the match.
+            state (State): The current state of the game.
+            action (Action): The action taken by the player.
+            reward (float): The reward received for the action.
+        """
+        if match_id not in self.matches:
+            self.matches[match_id] = Match(
+                match_id=match_id,
+                white_player=RandomPlayer(color=Color.WHITE),
```

Comment:
pass white and black player as parameter as AbstractPlayer and simply assign them to the match object